students/duanez2021/lesson05/mailroom_rev03-1.py

```diff
@@ -0,0 +1,224 @@
+# Changes:
+#   donor_db to a dictionary object (see mailroom_rev02-1.py)
+# Replaced:
+#   replaced these functions from from mailroom_rev02-1.py to use dictionary objects
+#   create_required_report
+#   added get_sum_donations
+#   get_number_of_donations
+#   get_avg_donation_amt
+#   check_donor_name
+#   list_donors
+#   add_new_donor
+#   resolve_donor_name
+#   add_donor_amt
+#   prep_individual_email
+#   send_email_to_all
+#
+#   additional changes (for lesson 5)
+#   add_donor_amt (added try-block to catch entry of inappropriate amount values)
+#   prep_individual_email (added try-block to catch error raised by add_donor_amt)
+#   send_email_to_all list (added list comprehension)
+#   main (added try-block to catch error raised by inappropriate menu value entries)
+#
+#   Duane McCollum Python self-paced winter 2020
+#
+#######################################################################################################
+# original donor db, was a list
+
+import sys, os  # imports go at the top of the file
+#print(os.getcwd())
+donor_db = {"William Gates, III": [653772.32, 12.17],
+            "Jeff Bezos": [877.33],
+            "Paul Allen": [663.23, 43.87, 1.32],
+            "Mark Zuckerberg": [1663.23, 4300.87, 10432.0],
+            "Catherine Zeta-Jones": [100.00, 100.00, 1000.00]
+            }
+
+new_prompt = "\n".join(("Choose an action:\n",
+                    "1 - Send a Thank You to a single donor.",
+                    "2 - Create a Report.",
+                    "3 - Send letters to all donors.",
+                    "4 - Quit\n",
+                    ">>> "))
+
+
+thankyou_2_prompt = "Send thank you to (full name)\n(Enter \'list\' for list of donors;\n\'exit\' to return to main menu): "
+thankyou_for_amt_prompt = ("Donation amount: ")
+
+def create_required_report():
+    print('{0: <20}{1} {2: >12} {3} {4: ^10} {5} {6: >13}'.format('Donor Name','|', 'Total Given','|', 'Num Gifts','|','Average Gift'))
+    print('{:''^66}'.format(' '))
+    for idx, donor_name in enumerate(donor_db):
+        t = (donor_name, get_sum_donations(donor_name), get_number_of_donations(donor_name), get_avg_donation_amt(donor_name))
+        strFormat = '{0: <20.20}${1: >13,.2f}  {2:^11}  ${3:> 13,.2f}'
+        print(strFormat.format(t[0], t[1], t[2], t[3]))
+    print('\n')
+    return True
+
+
+def get_sum_donations(donor_name):
+    sum_amt = sum(donor_db[donor_name])
+    return sum_amt
+
+def get_number_of_donations(donor_name):
+    n = len(donor_db[donor_name])
+    return n
+
+def get_avg_donation_amt(donor_name):
+    sm = get_sum_donations(donor_name)
+    n = get_number_of_donations(donor_name)
+    return round(sm/n, 2)
+
+#check_donor_name('Jeff Bezos')
+def check_donor_name(donor_name):
+    """this function returns True if the
+        donor name is found in the donor_db.
+    """
+    b = False
+    for idx, dname in enumerate(donor_db):
+        if donor_name.lower() == dname.lower():
+            b = True
+    return b
```

Comment:
this is not a correct way to loop up a key in the dictionary, use get() or [<key>] access.